File: patterns_prac/bt_dp_prac0/dp_grid_traveller.py
# this is dp based grid traveller

# Given a n x n grid starting at top left
# find how many ways you can reach the
# bottom right corner. You can move right & down


def gt(m, n):
    # Initialize a 2D DP table
    dp = [[0 for _ in range(n + 1)] for _ in range(m + 1)]

    # Base case
    dp[1][1] = 1

    # Fill the DP table
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if i == 1 and j == 1:
                continue  # Skip the base case
            dp[i][j] = (dp[i - 1][j] if i > 1 else 0) + (
                dp[i][j - 1] if j > 1 else 0
            )  # Recurrence relation
            # number of ways to reach the cell above curr
            # number of ways to reach the cell left of curr

    return dp[m][n]

File: patterns_prac/bt_dp_prac2/src/bin/dp_grid_traveller.rs
fn main() {
    println!("Grid Traveller:");
    println!("{}", grid_traveller(3, 3));
}

fn grid_traveller(m: usize, n: usize) -> usize {
    // how many ways can you reach bottom right
    // initialize a 2D dp table
    let mut dp = vec![vec![0; m + 1]; n + 1];

    // base case
    dp[1][1] = 1;

    // enumerate and fill in the DP table
    // with the recurrence relation
    for idx in 1..m + 1 {
        for jdx in 1..n + 1 {
            if jdx == 1 && idx == 1 {
                continue;
            }
            dp[idx][jdx] = (if idx > 1 { dp[idx - 1][jdx] } else { 0 })
                + (if jdx > 1 { dp[idx][jdx - 1] } else { 0 });
        }
    }
    dp[m][n]
}
